lib/auth/github.go

```diff
@@ -0,0 +1,425 @@
+		token.TokenType, token.Expires, token.Scope)
+	// Github does not support OIDC so we have to retrieve user information
+	// by making requests to its API using the access token we just got
+	claims, err := populateGithubClaims(&githubAPIClient{token: token.AccessToken})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if len(connector.GetTeamsToLogins()) != 0 {
+		err = s.createGithubUser(connector, *claims)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+	response := &GithubAuthResponse{
+		Identity: services.ExternalIdentity{
+			ConnectorID: connector.GetName(),
+			Username:    claims.Email,
+		},
+		Req: *req,
+	}
+	if !req.CheckUser {
+		return response, nil
+	}
+	user, err := s.Identity.GetUserByGithubIdentity(response.Identity)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	response.Username = user.GetName()
+	roles, err := services.FetchRoles(user.GetRoles(), s.Access, user.GetTraits())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if req.CreateWebSession {
+		session, err := s.NewWebSession(user.GetName())
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		sessionTTL := roles.AdjustSessionTTL(defaults.OAuth2TTL)
+		bearerTTL := utils.MinTTL(BearerTokenTTL, sessionTTL)
+		session.SetExpiryTime(s.clock.Now().UTC().Add(sessionTTL))
+		session.SetBearerTokenExpiryTime(s.clock.Now().UTC().Add(bearerTTL))
+		err = s.UpsertWebSession(user.GetName(), session)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		response.Session = session
+	}
+	if len(req.PublicKey) != 0 {
+		certTTL := utils.MinTTL(defaults.OAuth2TTL, req.CertTTL)
+		allowedLogins, err := roles.CheckLoginDuration(
+			roles.AdjustSessionTTL(certTTL))
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		cert, err := s.GenerateUserCert(
+			req.PublicKey,
+			user,
+			allowedLogins,
+			certTTL,
+			roles.CanForwardAgents(),
+			roles.CanPortForward(),
+			req.Compatibility)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		response.Cert = cert
+		authorities, err := s.GetCertAuthorities(services.HostCA, false)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		for _, authority := range authorities {
+			response.HostSigners = append(response.HostSigners, authority)
+		}
+	}
+	s.EmitAuditEvent(events.UserLoginEvent, events.EventFields{
+		events.EventUser:   user.GetName(),
+		events.LoginMethod: events.LoginMethodGithub,
+	})
+	return response, nil
+}
+
+func (s *AuthServer) createGithubUser(connector services.GithubConnector, claims services.GithubClaims) error {
+	logins := connector.MapClaims(claims)
+	log.WithFields(log.Fields{trace.Component: "github"}).Debugf(
+		"Generating dynamic identity %v/%v with logins: %v",
+		connector.GetName(), claims.Email, logins)
+	user, err := services.GetUserMarshaler().GenerateUser(&services.UserV2{
+		Kind:    services.KindUser,
+		Version: services.V2,
+		Metadata: services.Metadata{
+			Name:      claims.Email,
```

Comment:
I am pretty sure email is mandatory for every Github account. We can use a username instead I think, should not be an issue.

Reply:
Only in this case we'd have to request the scope `user` which is broader than `user:email` which we request now (and which grants permissions to read only email addresses).

Reply:
I've updated the code to use Github username instead of email. Actually username is public information so we don't need any additional scopes for it, so we only request `read:org` now.